### DISCUSSION TITLE
Possibility to keep the realtime connections open?

### DISCUSSION THREAD

**Discussion by aluedtke7** [pocketbase/pocketbase#1869] [Q&A] [answered]
I host my pocketbase on fly.io and my Flutter App is relying on the realtime connections. Unfortunately, fly is closing these connections when they are idle for more than 60 seconds.

Ist there anything I can do in my app to keep them open? Like a ping every 50s or so.

**Comment by ganigeorgiev** [comment, marked as answer]
At the moment there is no integrated option for client ping.

If you are extending PocketBase, then you can bind to the [`OnRealtimeConnectRequest` hook](https://pocketbase.io/docs/event-hooks/#realtime-api-hooks) and use a timer/ticker, `time.AfterFunc`, etc. and manually send a ping message to the client. Something like:

```go
package main

import (
	"log"
	"time"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/subscriptions"
)

func ping(client subscriptions.Client, interval time.Duration) {
	time.AfterFunc(interval, func() {
		if client == nil || client.IsDiscarded() {
			return
		}

		msg := subscriptions.Message{
			Name: "ping",
		}

		client.Channel() <- msg

		ping(client, interval)
	})
}

func main() {
	app := pocketbase.New()

	app.OnRealtimeConnectRequest().Add(func(e *core.RealtimeConnectEvent) error {
		ping(e.Client, 30*time.Second)

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

_But note that you may not need the above because the SDKs (and any SSE client in general) will handle the reconnect for you._

**Reply by aluedtke7** [reply]
Thank you for the detailed answer.
I've made the observation, that my app (Flutter on Android) sometimes doesn't get the realtime events. This might be due to the poor internet connection in some shops.

When I test the same (wake up of app after several minutes of inactivity) in my local WiFi, the events always arrive in my app.
So I agree with you, that the reconnect mechanism does work as expected when the internet connection is good. Maybe there is room for little improvements (try to resend the event more often?) when the connection is bad.

**Reply by ganigeorgiev** [reply]
Yes, the Dart SDK realtime handling could be improved. Currently it has a 5s reconnect timeout for max 5 attempts, but as you noted this may not always work, especially if the app is inactive or has no internet access for several minutes. The reconnect attempts shouldn't be limited (unless the server previously disallowed it) and the timeout should be shorter.

I'll consider updating the reconnect handling sometime next week to unify it with the JS SDK (_there the reconnect is "stepped", aka. it will try first 200ms, 300ms, 500ms, 1000ms, 1200ms, 1500ms, 2000ms and then will repeat indefinitely until the sse client is active only with the last value - 2000ms_).

**Reply by aluedtke7** [reply]
Thank you. I'm looking forward for that!

**Reply by ganigeorgiev** [reply]
_Just for info, the autoreconnect Dart SDK handling was updated in [v0.7.3](https://pub.dev/packages/pocketbase/changelog#073).
Also please note that currently the realtime events are not buffered, meaning that the client will not receive the events that have happened during the time the app wasn't active or didn't have internet access. If you want to handle this case, you have to manually refetch the items list on app/device wakeup and/or internet connectivity change._
